tests/src/main/java/net/neoforged/neoforge/debug/transfer/VanillaHandlersTests.java

```diff
@@ -0,0 +1,566 @@
+        var furnaceWrapper = VanillaContainerWrapper.of(furnace);
+
+        IntSupplier cookingTimerAccessor = () -> ObfuscationReflectionHelper.getPrivateValue(AbstractFurnaceBlockEntity.class, furnace, "cookingTimer");
+
+        helper.runAtTickTime(5, () -> {
+            helper.assertTrue(cookingTimerAccessor.getAsInt() > 0, "Furnace should have started cooking.");
+
+            try (Transaction transaction = Transaction.open(null)) {
+                if (furnaceWrapper.extract(rawIron, 64, transaction) != 64) {
+                    throw helper.assertionException("Failed to extract 64 raw iron.");
+                }
+            }
+
+            helper.assertTrue(cookingTimerAccessor.getAsInt() > 0, "Furnace should still cook after simulation.");
+
+            try (Transaction transaction = Transaction.open(null)) {
+                if (furnaceWrapper.extract(rawIron, 64, transaction) != 64) {
+                    throw helper.assertionException("Failed to extract 64 raw iron.");
+                }
+                // Even if we re-insert, the furnace got emptied at some point so it should reset its cook timer
+                if (furnaceWrapper.insert(rawIron, 64, transaction) != 64) {
+                    throw helper.assertionException("Failed to extract 64 raw iron.");
+                }
+
+                transaction.commit();
+            }
+
+            helper.assertTrue(cookingTimerAccessor.getAsInt() == 0, "Furnace should have reset cook time after being emptied.");
+
+            helper.succeed();
+        });
+    }
+
+    /**
+     * Tests that the passed block doesn't update adjacent comparators until the very end of a committed transaction.
+     *
+     * @param block    A block with a Container block entity.
+     * @param resource The resource to try to insert (needs to be supported by the Container).
+     */
+    private static <T extends BlockEntity & Container> void testComparatorOnContainer(ExtendedGameTestHelper helper, Block block, ItemResource resource, Class<T> containerClass) {
+        var level = helper.getLevel();
+
+        BlockPos pos = new BlockPos(0, 1, 0);
+        helper.setBlock(pos, block.defaultBlockState());
+        T container = helper.getBlockEntity(pos, containerClass);
+        var resourceHandler = VanillaContainerWrapper.of(container);
+
+        BlockPos comparatorPos = new BlockPos(1, 1, 0);
+        Direction comparatorFacing = helper.getTestRotation().rotate(Direction.WEST);
+        // support block under the comparator
+        helper.setBlock(comparatorPos.relative(Direction.DOWN), Blocks.GREEN_WOOL.defaultBlockState());
+        // comparator
+        helper.setBlock(comparatorPos, Blocks.COMPARATOR.defaultBlockState().setValue(ComparatorBlock.FACING, comparatorFacing));
+
+        try (Transaction transaction = Transaction.open(null)) {
+            if (level.getBlockTicks().hasScheduledTick(helper.absolutePos(comparatorPos), Blocks.COMPARATOR)) {
```

Comment:
Can we just store the absolute comparator pos, instead of inlining in a bunch of places converting it from relative to absolute?